### ISSUE TITLE
Trait objects can call nonexistent concrete methods

### ISSUE THREAD

**Issue by matthewjasper** [rust-lang/rust#50781]
This code segfaults at runtime

```rust
use std::{
    fmt::{self, Debug},
};

trait X {
    fn foo(&self) where Self: Debug;
}

impl X for fn(&()) {
    fn foo(&self) where fn(&()): Debug {}
}

impl Debug for dyn X {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "hi!")
    }
}

pub fn main() {
    let x: fn(&()) = |_| ();
    let y = &x as &dyn X;
    y.foo(); // Segfault at opt-level 0, SIGILL otherwise.
}
```

cc #48557 - Makes this easier to come across.

**Comment by leodasvacas** [comment]
I did some investigation. The ideal fix is to catch this at declaration site, in the well-formedness (WF) check. Failing that we should be catching it at the call site, but that code dodged all checks.

The presence of a `for<'a> fn(&'a ())`, the desugaring of `fn(&())`, brings suspicion to uses of `skip_binder()` which would skip the binder `for<'a>` and turn the `'a` into what we call an "escaping region".

The WF check for the `for<'a> fn(&'a ()): Debug` trait predicate is reached by a call to `check_where_clauses`, then to `ty::wf::predicate_obligations` where we confidently skip the binder
https://github.com/rust-lang/rust/blob/2a3f5367a23a769a068c37460db336de427c4b48/src/librustc/ty/wf.rs#L75-L78 
only to completely ignore any trait predicates containing an escaping region in `compute_trait_ref`, so the predicate is not checked at all for WF.
https://github.com/rust-lang/rust/blob/2a3f5367a23a769a068c37460db336de427c4b48/src/librustc/ty/wf.rs#L191-L196

I don't know how to fix this but this certainly looks like guilty code.

**Comment by nikomatsakis** [comment]
OK, so, this first of all is not dependent on #48557. For example, this example will reproduce the problem on stable:

https://play.rust-lang.org/?gist=8c8b887b74da66782339e3a79c21c378&version=stable&mode=debug

The problem is rather a flaw in the object safety rules. Specifically, the object safety rules are supposed to prevent `Self` from appearing in the where clauses:

https://github.com/rust-lang/rust/blob/dbd10f81758381339f98994b8d31814cf5e98707/src/librustc/traits/object_safety.rs#L132-L134

But for some reason we seem to permit `Self` in the special case of `Self: Trait`:

https://github.com/rust-lang/rust/blob/dbd10f81758381339f98994b8d31814cf5e98707/src/librustc/traits/object_safety.rs#L167-L170

I'm actually..not sure why we do that. It seems ok on the trait level, as that is validated at the impl site, but not at the method level.

**Comment by nikomatsakis** [comment]
Marking as P-high since the fix may break code.

**Comment by nikomatsakis** [comment]
So I was discussing this problem with myself (I don't think @leodasvacas is around =) on Discord. You can search for this GUID to find it: 3b75dc47-42f1-44da-a39e-562f5875e932

However, the TL;DR is that we don't have to forbid *all* where clauses like `Self: B` where `B` is some bound. The key question is whether proving that `dyn Trait: B` *implies* that the hidden `Self` also satisfies `B`. This is true for lifetime bounds but also auto traits. It is not true for general traits (which can be implemented for a `dyn Trait` type but not necessarily other types). 

So I think we should loosen the conditions in the PR, which should help with regressions.

**Comment by nikomatsakis** [comment]
Nominating for lang team discussion.

**Comment by leodasvacas** [comment]
Fwiw the `typemap` crate, which was the major regression caused by the proposed patch, can be [hacked into working](https://github.com/reem/rust-typemap/pull/41) with the rule that `Self: Send` makes the method not object-safe.

I think that proposal 2 would be best from a language point of view, it sounds like bad design in principle, but the problematic situations never happen in practice so it would be a an edge case fix for an edge case problem. However it may be impractical to implement, I don't think the compiler can readily tell us if an impl that looks like `impl Bar for dyn Foo + ..` exists or not.

**Comment by daboross** [comment]
Forgive me if I'm not fully understanding this issue, but would it be a good solution to have `Self: AnyOtherTrait` on a method just also imply `Self: Sized`? So methods marked with `where Self: Debug` or any other trait just wouldn't be callable on dynamic objects. This seems like it would be a sane solution that seemly wouldn't break existing sound code.

Being able to `impl Trait for OtherTrait` is a very useful feature, especially when dealing with `Any` and similar dynamic tricks. It would seem very strange to remove this for the sake of an edge-case in `where` in methods.

**Comment by nikomatsakis** [comment]
> Fwiw the typemap crate, which was the major regression caused by the proposed patch, can be hacked into working with the rule that Self: Send makes the method not object-safe.

Good to know, though I still feel like the original code was "reasonable" somehow.

> However it may be impractical to implement, I don't think the compiler can readily tell us if an impl that looks like impl Bar for dyn Foo + .. exists or not.

It'd be a bit tricky, but I think it's *possible* — in particular, such an impl would have to be in the crate where the trait `Foo` is located (which may or may not be the same crate where `Bar` is located).

---

@daboross 

> Forgive me if I'm not fully understanding this issue, but would it be a good solution to have

That's certainly an option — it has the downside though that you would not be able to call such methods via trait objects, even when -- like with `Send` -- it seems reasonable to do so.

**Comment by leodasvacas** [comment]
> in particular, such an impl would have to be in the crate where the trait Foo is located

I'm not so sure, I recently learned that the following is coherent:
```rust
// Crate A
pub trait X<T> {
    fn foo(&self) where Self: PartialEq<T>;
}

// Crate B
struct Mine;

impl X<Mine> for Mine {
    fn foo(&self) {}
}

impl PartialEq<Mine> for X<Mine> { fn eq(&self, _: &Mine) -> bool { true } }

pub fn main() {
    <X<Mine> as X<Mine>>::foo(&Mine); // Segfault at opt-level 0, SIGILL otherwise.
}
```

About @daboross suggestion, I think the reason we don't generally imply `Self: Sized` is because we'd like to remove many of the current restrictions on trait objects in the future. If we start implying `Self: Sized` for a certain signature then we have to imply it forever, even if object safety rules are relaxed.

**Comment by nikomatsakis** [comment]
So branching off from [this comment](https://github.com/rust-lang/rust/issues/51443#issuecomment-400996835) on #51443, let's try to see if we can state precisely the rules that we are happy with. As I wrote there, the rule we had in mind is **roughly** this:

- If a trait `Trait` has a method `method` with a where-clause `where Self: XXX`
  - that is object-safe so long as `dyn Trait: XXX` cannot be proven

This precise statement is a bit weird in the case where `trait Trait: XXX` -- we could perhaps try to distinguish those cases; in Chalk terms, `XXX` would be an implied bound of `Trait`.

At some point I was considering this as a generalization of the rules around `where Self: Sized`, but I realize now that while they are related, they are not the same. In that case, we know that `dyn Trait: Sized` could *never* be proven, and hence we can suspend the where-clause requirements entirely.

**Comment by leodasvacas** [comment]
A perhaps better alternative is to check the cast sites rather than the trait, so given a trait definition:
```rust
pub trait X<T> {
    fn foo(&self) where Self: Other<T>, T: Other<Self>
}
```
When doing a cast like `U as DynTy` where `DynTy = dyn X<Foo> + ...` then for the predicates `Self: Other<Foo>` and `Foo: Other<Self>` we'd check that if they hold when `Self` is substituted by `DynTy` they then must hold if `Self` is substituted by `U`:
```rust
Holds(DynTy: Other<Foo>) => Holds(U: Other<Foo>)
Holds(Foo: Other<DynTy>) => Holds(Foo: Other<U>)
```
This should guarantee that any methods available for the dyn type are also available for the type that was erased.

**Comment by lahwran** [comment]
I'm going through the `I-unsound` issues to review their current status and next steps. Looks like the current status of this issue is that the compiler emits a warning for the entire class of unsoundness, leaving no holes arising from this issue in a program that compiles without warning, and further progress is expected to be possible once chalk is ready. If I'm wrong that the warning covers all potential instances of this dyn missing method issue, I'd love to be corrected. Seems like consensus is that this isn't viable to properly fix in rustc until chalk is ready.

**Comment by oli-obk** [comment]
```rust
trait Trait {}

trait X {
    fn foo(&self) where Self: Trait;
}

impl X for () {
    fn foo(&self) {}
}

impl Trait for dyn X {}

pub fn main() {
    <X as X>::foo(&()); // Segfault at opt-level 0, SIGILL otherwise.
}
```

still segfaults and miri says

```
error: Undefined Behavior: `dyn` call trying to call something that is not a method
  --> src/main.rs:14:5
   |
14 |     <dyn X as X>::foo(&()); // Segfault at opt-level 0, SIGILL otherwise.
   |     ^^^^^^^^^^^^^^^^^^^^^^ `dyn` call trying to call something that is not a method
```

**Comment by Jules-Bertholet** [comment]
There's one potential (partial) fix that I haven't seen discussed, that would continue to support many of the use-cases people have for the trait bounds in question. Notably, it would allow `async-trait` to keep working without the additional restrictions on auto-trait impls that #107082 imposed.


Let's look at the MCVE for this issue:

```rust
trait Trait {}

trait X {
    fn foo(&self) where Self: Trait;
}

impl X for () {
    fn foo(&self) {}
}

impl Trait for dyn X {}

pub fn main() {
    <dyn X as X>::foo(&()); // Segfault at opt-level 0, SIGILL otherwise.
}
```

The source of the problem is that there is no source for the implementation of `<dyn X as X>::foo`. Now, let's consider a very similar example:

```rust
trait Trait {}

trait X {
    // `foo` now has a default impl
    fn foo(&self) where Self: Trait {}
}

impl X for () {}

impl Trait for dyn X {}

pub fn main() {
    <dyn X as X>::foo(&()); // Segfault at opt-level 0, SIGILL otherwise.
}
```

In this modified example, there *is* a potential source for the `<dyn X as X>::foo` implementation: the default impl! And in fact, given the current state of stable Rust (#20021, no `#![feature(trivial_bounds)]`), the common case is likely to be that a default method impl exists. And the discussion in #51443 seems to confirm this. So, a potential fix to this issue would be to use the default impl to provide the `dyn Trait` impl whenever possible, and impose restrictions on object safety only when no default is present.

One wrinkle is that additional rules would be necessary to prevent lifetime-based specialization. Consider:

```rust
#![feature(trivial_bounds)]

trait Trait {}

impl Trait for &'static () {}

trait X {
    // `foo` now has a default impl
    fn foo(&self) where Self: Trait {}
}

impl<'a> X for &'a () {
    fn foo(&self) where Self: Trait {
        println!("`Self: 'static` must hold if this is ever called");
    }
}

impl<'a> Trait for dyn X + 'a {}

pub fn test<'a>(r: &'a ()) {
    <dyn X + 'a as X>::foo(&r); // Segfault at opt-level 0, SIGILL otherwise.
}
```

In this case, the implementation of `<dyn X + 'a as X>::foo` would have to be taken from the default *even in the case of `'a: 'static`*. Taking it from the `impl<'a> X for &'a ()` block would be unsound for `'a` ≠ `'static`, and choosing a different impl depending `'a` would be unsound lifetime-dependent specialization.

**Comment by fmease** [comment]
Fixed by #125380.
